### ISSUE TITLE
Support caching RPMs for test environments

### ISSUE THREAD

**Issue by jlebon** [jlebon/redhat-ci#36]
A *lot* of the testers do `dnf install -y <big list of deps>`. We should look into somehow caching the RPMs "locally" and e.g. injecting repo files in the environments?

But how to know which RPMs to cache without forcing everyone to use a proxy? Maybe we can just parse the `packages` stanzas, resolve them, and cache those only.

**Comment by cgwalters** [comment]
I think to start, for every VM we provision, override the `/etc/yum.repos.d` to point to a local caching proxy.  It's a bit hard to do this for `docker run` but that's something that would be really useful to implement generally.

**Comment by cgwalters** [comment]
We also can easily do this for provisioned containers too, the hard case is the "docker run in VM".  (Not to mention in Kube/OpenShift).

**Comment by cevich** [comment]
If only Red Hat had [a product that was designed for caching rpms](https://access.redhat.com/products/red-hat-update-infrastructure/)...and some sort of ever present "slave" system that always lives nearby the test VMs 😀

**Comment by cevich** [comment]
Ugg, nvm, shouda RTFM first, says you need 3vms minimum for RHUI.  Though, I still bet you could use squid and just shut off all the https stuff in yum.  That would make setup a lot easier.

**Comment by jlebon** [comment]
Hmm, I'm still not sure about using a proxy for this. It wouldn't be able to cache HTTPS, so we'd have to make sure HTTP was used. Even then, mirrors will make proxy caching even less effective. So we would have to do something like: for each repo in `/etc/yum.repos.d/`, check that it's gpg enabled, and if so, then drop any metalink/mirrorlist entries and inject the canonical baseurl (which we'd have to hardcode somewhere). It's doable, but yucky.

I'm thinking now the saner (and less intrusive) thing to do is to use the shotgun approach and just mirror selected repos, e.g. `fedora` and `fedora-updates` and inject that with a `cost` less than 1000.

```
$ curl -s http://dl.fedoraproject.org/pub/DIRECTORY_SIZES.txt | grep 'pub/fedora/linux/releases/25/Everything/x86_64/os$'                                                                                                   
55G     /pub/fedora/linux/releases/25/Everything/x86_64/os
$ curl -s http://dl.fedoraproject.org/pub/DIRECTORY_SIZES.txt | grep 'pub/fedora/linux/updates/25/x86_64$'
51G     /pub/fedora/linux/updates/25/x86_64
```

Given the bandwidth we'd save, I say it's definitely worth the 106G. Not to mention this is a really useful service that has use cases outside RHCI.

**Comment by cevich** [comment]
Getting yum to use http isn't that big of a deal, but between that and also setting up squid, it's certainly adding a more things to go wrong.  OTOH, if your mirror ever goes missing / inaccessible, it means ALL the hosts coded to use it will fail.  I think using the internal CDN mirrors are a good compromise between network-locality and reliability.

Another idea is to constantly yum update some image and snapshot it.  Then provision all the hosts from the snapshot (I'm assuming creating the snapshot is atomic for in-flight creations).

**Comment by jlebon** [comment]
The mirror would be *in addition* to the standard repo files, so even if it's down it should fallback to the usual ones. Specifying a cost less than 1000 should make yum prefer it over other mirrors given the same NEVRA.

That said, I was reading up on the Fedora infra, and I'm surprised we don't have an internal private mirror hooked up to MirrorManager2. Might open a ticket for that. Until then, it wouldn't be very hard to keep a repo synced for our purposes.

**Comment by jlebon** [comment]
https://github.com/jlebon/redhat-ci/pull/40

**Comment by jlebon** [comment]
Closed by #40.
